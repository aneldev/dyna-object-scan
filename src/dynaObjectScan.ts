export const dynaObjectScan = (
  obj: any,
  cb: (args: {
    value: any;
    parent: any;
    propertyName?: string;
    path?: string;
    skip: () => void;
  }) => void,
  _parent?: any,
  _propertyName?: string,
  _path: string = '',
  _scanned: any = [],
): void => {
  if (typeof obj === 'object' && obj !== null && _scanned.indexOf(obj) > -1) return;
  _scanned.push(obj);

  let skip = false;

  cb({
    value: obj,
    parent: _parent,
    propertyName: _propertyName,
    path: _path,
    skip: () => skip = true,
  });

  if (skip) return;

  if (typeof obj === "object" && obj !== null) {
    if (Array.isArray(obj)) {
      obj
        .forEach((itemValue, index) => dynaObjectScan(
          itemValue,
          cb,
          obj,
          index.toString(),
          `${_path}[${index.toString()}]`,
          _scanned,
        ));
    }
    else {
      Object.keys(obj)
        .forEach(propertyName => dynaObjectScan(
          obj[propertyName],
          cb,
          obj,
          propertyName,
          `${_path}.${propertyName}`,
          _scanned,
        ));
    }
  }
};

