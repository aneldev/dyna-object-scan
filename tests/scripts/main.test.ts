import "jest";

jasmine.DEFAULT_TIMEOUT_INTERVAL = 4000;

import {dynaObjectScan} from "../../src";

// help: https://facebook.github.io/jest/docs/expect.html

describe("dynaObjectScan", () => {
  test('Scan simple object', () => {
    const output: any[] = [];
    dynaObjectScan(
      {
        fname: 'John',
        lname: 'Smith',
        age: 37,
      },
      ({path, value}) => output.push({path, value}),
    );
    expect(output).toMatchSnapshot();
  });

  test('Scan object with nested objects', () => {
    const output: any[] = [];
    dynaObjectScan(
      {
        fname: 'John',
        lname: 'Smith',
        age: 37,
        favoriteSweets: {
          panCakes: true,
          strawberries: false,
          kiwi: 'maybe',
        },
      },
      ({path, value}) => output.push({path, value}),
    );
    expect(output).toMatchSnapshot();
  });

  test('Scan object with nested objects and arrays', () => {
    const output: any[] = [];
    dynaObjectScan(
      {
        fname: 'John',
        lname: 'Smith',
        age: 37,
        favoriteSweets: {
          panCakes: true,
          strawberries: false,
          kiwi: 'maybe',
        },
        favoriteCars: 'Honda,Porsche'.split(','),
      },
      ({path, value}) => output.push({path, value}),
    );
    expect(output).toMatchSnapshot();
  });

  test('Scan object and skip', () => {
    const output: any[] = [];
    dynaObjectScan(
      {
        fname: 'John',
        lname: 'Smith',
        age: 37,
        favoriteSweets: {
          panCakes: true,
          strawberries: false,
          kiwi: 'maybe',
        },
        favoriteCars: 'Honda,Porsche'.split(','),
      },
      ({path, value, skip, propertyName}) => {
        if (propertyName === 'favoriteSweets') {
          skip(); // This will block the scan of the `favoriteSweets` property
          return;
        }
        output.push({path, value})
      },
    );
    expect(output).toMatchSnapshot();
  });

  test('Scan array', () => {
    const output: any[] = [];
    dynaObjectScan(
      [
        'Apple',
        'Banana',
        {
          name: 'Lola',
          age: 32,
        },
      ],
      ({path, value}) => output.push({path, value}),
    );
    expect(output).toMatchSnapshot();
  });
});
