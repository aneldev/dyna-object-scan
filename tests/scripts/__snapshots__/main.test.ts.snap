// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dynaObjectScan Scan array 1`] = `
Array [
  Object {
    "path": "[0]",
    "value": "Apple",
  },
  Object {
    "path": "[1]",
    "value": "Banana",
  },
  Object {
    "path": "[2]",
    "value": Object {
      "age": 32,
      "name": "Lola",
    },
  },
  Object {
    "path": "[2].name",
    "value": "Lola",
  },
  Object {
    "path": "[2].age",
    "value": 32,
  },
]
`;

exports[`dynaObjectScan Scan object and skip 1`] = `
Array [
  Object {
    "path": ".fname",
    "value": "John",
  },
  Object {
    "path": ".lname",
    "value": "Smith",
  },
  Object {
    "path": ".age",
    "value": 37,
  },
  Object {
    "path": ".favoriteCars",
    "value": Array [
      "Honda",
      "Porsche",
    ],
  },
  Object {
    "path": ".favoriteCars[0]",
    "value": "Honda",
  },
  Object {
    "path": ".favoriteCars[1]",
    "value": "Porsche",
  },
]
`;

exports[`dynaObjectScan Scan object with nested objects 1`] = `
Array [
  Object {
    "path": ".fname",
    "value": "John",
  },
  Object {
    "path": ".lname",
    "value": "Smith",
  },
  Object {
    "path": ".age",
    "value": 37,
  },
  Object {
    "path": ".favoriteSweets",
    "value": Object {
      "kiwi": "maybe",
      "panCakes": true,
      "strawberries": false,
    },
  },
  Object {
    "path": ".favoriteSweets.panCakes",
    "value": true,
  },
  Object {
    "path": ".favoriteSweets.strawberries",
    "value": false,
  },
  Object {
    "path": ".favoriteSweets.kiwi",
    "value": "maybe",
  },
]
`;

exports[`dynaObjectScan Scan object with nested objects and arrays 1`] = `
Array [
  Object {
    "path": ".fname",
    "value": "John",
  },
  Object {
    "path": ".lname",
    "value": "Smith",
  },
  Object {
    "path": ".age",
    "value": 37,
  },
  Object {
    "path": ".favoriteSweets",
    "value": Object {
      "kiwi": "maybe",
      "panCakes": true,
      "strawberries": false,
    },
  },
  Object {
    "path": ".favoriteSweets.panCakes",
    "value": true,
  },
  Object {
    "path": ".favoriteSweets.strawberries",
    "value": false,
  },
  Object {
    "path": ".favoriteSweets.kiwi",
    "value": "maybe",
  },
  Object {
    "path": ".favoriteCars",
    "value": Array [
      "Honda",
      "Porsche",
    ],
  },
  Object {
    "path": ".favoriteCars[0]",
    "value": "Honda",
  },
  Object {
    "path": ".favoriteCars[1]",
    "value": "Porsche",
  },
]
`;

exports[`dynaObjectScan Scan simple object 1`] = `
Array [
  Object {
    "path": ".fname",
    "value": "John",
  },
  Object {
    "path": ".lname",
    "value": "Smith",
  },
  Object {
    "path": ".age",
    "value": 37,
  },
]
`;
